Minimal types setup

dapplion opened the pull request:
Bare minimum setup to host and check types.
- package.json: declares typescript version used. Can be installed locally by running `npm install` or `yarn`
- tsconfig.json: declares ECMAScript version used (javascript version) for the Typescript compiler.

To make sure types compile run 
```
./node_modules/.bin/tsc
```

dot-asm commented:
Just a quick not to say **thanks** and give Season's Greetings:-) When I said "no rush" I meant "don't rush":-) In other words I'll look into it a little bit later. Thanks again and cheers.

dot-asm reviewed (commented):
How do you feel about naming `bindings.ts` as `blst.ts` instead?

dot-asm commented on bindings/node.js/package.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "devDependencies": {
+    "@types/node": "^14.14.14",
+    "typescript": "^4.0.3"
```
I'm prefer-actual-minimal-requirement-instead-of-whatever-latest-for-the-moment kind of guy, and I had no problem with 3.8.3. In other words "^3" would supposedly do. However...

I'm also trying to keep things at bare minimum on my desktop, and when I tried the `npm install` as you suggested, it resulted in errors. After which I've figured that if I start from scratch and go straight for `tsc`, the one I have on `$PATH`, I get a meaningful message. It suggested to run `npm i @types/node`, after which it succeeded. And it doesn't matter if I have package.json **or not**. In other words it appears that at least for intended purpose (of testing things prior pinging blst-ts for feedback and integration) this file appears to be superfluous. Well, it doesn't prevent me from running `npm install --global typescript@3` or `npm install @types/node@10`...

dapplion replied:
It's not strictly necessary now, but if you want to distribute this package through NPM you are required to have a package.json.

dot-asm commented on bindings/node.js/tsconfig.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "include": ["*"],
```
As matter of general education. Does `"*"` imply "don't pay attention to irrelevant files"? In other words is `"*"` effectively equal to `"*.ts"`? And in this specific context, would it be sufficient to include just "runnable.ts"?

dapplion replied:
`"*"` tells the Typescript compiler to consider all local files for compilation. In normal projects, one usually targets `src` which contains an `index.ts` file.

dot-asm commented on bindings/node.js/runnable.ts:
```diff
@@ -0,0 +1,42 @@
+const pk = new blst.P1(SK);
+const pk_for_wire = pk.serialize();
+
+const sig = new blst.P2();
```
As a general remark. While declaring `sig`, being effectively a pointer, `const` would supposedly work, `hash_to` and `sign_with` do modify the internal value of the object. In sense that calls to serialize would yield different values after each step.

dapplion replied:
The `const` here just refers to the pointer to the P2 instance. It does not prevent the instance from mutating itself.

dot-asm commented:
Just in case, it's not like I demand changes to be made in this PR. It's sufficient that we discuss and come to conclusions. After which I can make adjustments myself. If you wish that is:-)

dot-asm commented on bindings/node.js/bindings.ts:
```diff
@@ -0,0 +1,186 @@
+type bool = boolean;
+type size_t = number;
+type app__string_view = string | Uint8Array;
+type DST = string;
```
Some further reading shows that suggestion to use capitalized types is not appropriate in the context. It's suggested that `boolean`, `number`, `string` are true TS basic types. As well as `bigint` that will be needed now. Apparently `Uint8Array` falls out of the pattern and there is no corresponding "basic" type. Oh well, sorry about distraction. Question about `DST` still stands though...

Yet, as for Uint8Array. Since `type byte[s] = Uint8Array` is suggested, why not use `string | byte[s]`?

dapplion replied:
I initially used types that mimic the C types for ease of maintenance by C devs, but it's best to just follow the Typescript standard. We don't need the DST type indeed. Uint8Array is not strictly a basic type but could be considered as such. It's the cross-platform standard to represent bytes

dot-asm commented on bindings/node.js/bindings.ts:
```diff
@@ -0,0 +1,186 @@
+type size_t = number;
+type app__string_view = string | Uint8Array;
+type DST = string;
+type byte = Uint8Array;
```
Ideally one would want fixed-size arrays in most cases where the type is used, but it doesn't seem to be something one can pull off (with style:-). Would it be appropriate to embed lengths in parameter names, e.g. `from_bendian(_32: bytes): void;`?

dapplion replied:
Unfortunately, Typescript has no way to enforce specific byte lengths on byte arrays, eventually, it all is Uint8Array. So it's best to just use that and add a JSDoc comment with the expected byte length.

dot-asm commented:
I've committed *a* version of this. I've chosen to rename `bindings.ts` to `blst.hpp.ts`, added missing methods, amended declarations to reflect recent blst.swg changes, and added a trivial sanity check to ci.yml. Thanks!

dapplion commented on bindings/node.js/bindings.ts:
```diff
@@ -0,0 +1,186 @@
+    hash_or_encode: bool,
+    msg: app__string_view,
+    DST?: DST,
+    aug?: app__string_view
```
Uint8Array should be used here too

dapplion commented:
> I've committed _a_ version of this. I've chosen to rename `bindings.ts` to `blst.hpp.ts`, added missing methods, amended declarations to reflect recent blst.swg changes, and added a trivial sanity check to ci.yml. Thanks!

This PR totally slipped through the cracks. Thanks for picking it up! I've replied to all your comments, happy to answer more here or in a subsequent PR. The key question for you is if you want to distribute the library through NPM or not.
- (a) Distributing through NPM: That is how virtually all the Typescript + Javascript projects consume code. It's the sole purpose of https://github.com/ChainSafe/blst-ts which is then consumed by the Lodestar client through the https://github.com/ChainSafe/bls wrapper which exposes a WASM implementation too.
- (b) Do not distribute through NPM: Then (almost) no-one will be able to consume the Typescript code you committed. It can be used as a reference for other implementations but not much else.

Let me know your thoughts

dot-asm commented:
> * (b) Do not distribute through NPM: Then (almost) no-one will be able to consume the Typescript code you committed. It can be used as a reference for other implementations but not much else.

Current goal is specifically that, to provide a template. And "two-way" one, in sense that it's not only for applications to take, but even to make suggestions :-)

dapplion commented:
> > * (b) Do not distribute through NPM: Then (almost) no-one will be able to consume the Typescript code you committed. It can be used as a reference for other implementations but not much else.
> 
> Current goal is specifically that, to provide a template. And "two-way" one, in sense that it's not only for applications to take, but even to make suggestions :-)

Got it! Then the current setup works great for that. Let me know if I can contribute to anything else JS/TS related
